Refactor LowLightBoost event to state

temcguir opened the pull request:
This removes LLB "events" which can be more simply represented as a state within the existing LowLightBoostState.

This change merges the failure event into LowLightBoostState as a new 'Error' state. This allows for the removal of the lowLightBoostEvents channel from the CameraSystem interface, simplifying the public API and ensuring unidirectional data flow.

The PreviewViewModel and associated UI have been updated to handle the new `Error` state correctly.

This also renames the "Strength" LowLightBoostState to "Active" and adds an "Inactive" state to the sealed interface.

temcguir committed:
Refactor LowLightBoost event to state

This removes LLB "events" which can be more simply represented as
a state within the existing LowLightBoostState.

This change merges the failure event into LowLightBoostState as a new
'Error' state. This allows for the removal of the lowLightBoostEvents
channel from the CameraSystem interface, simplifying the public API
and ensuring unidirectional data flow.

The PreviewViewModel and associated UI have been updated to handle the
new `Error` state correctly.

donovanfm commented:
1) There's another call to `onLowLightBoostErrorCallback()` in `LowLightBoostSurfaceProcessor` in the `onSessionDisconnected()` callback that needs to have an Exception passed in now.
2) There's a behavior change here where errors are treated as non-transient. Maybe that's how we want it to behave as default. But if `onLowLightBoostErrorCallback()` and the FlashMode is switched to OFF, and then you cycle back through the FlashModes to LLB, it still has the Error state and automatically switches back to OFF until you reboot JCA.

temcguir committed:
Merge branch 'gllb' into temcguir/gllb_replace_events_with_state

temcguir committed:
Add exception to error callback for LLB

temcguir commented:
> 1. There's another call to `onLowLightBoostErrorCallback()` in `LowLightBoostSurfaceProcessor` in the `onSessionDisconnected()` callback that needs to have an Exception passed in now.

Added an exception to that new callback.

> 2. There's a behavior change here where errors are treated as non-transient. Maybe that's how we want it to behave as default. But if `onLowLightBoostErrorCallback()` and the FlashMode is switched to OFF, and then you cycle back through the FlashModes to LLB, it still has the Error state and automatically switches back to OFF until you reboot JCA.

If you're seeing that behavior, how are you reproducing it? The collectors on the flow will only set the flash mode to OFF when a new LowLightBoostState.Error is emitted. Are you sure you're not just receiving errors every time you try to switch to LLB?

donovanfm commented:
For testing this, I'm throwing an Exception every time in `tryCreateLowLightBoostSession()`. But before this change it would show the error message Snackbar each time you tried to enable LLB, and with this change it only shows the Snackbar on the first time. I can do some more testing to confirm.

temcguir committed:
Only handle LLB state change errors when the state changes

 Adds distinctUntilChanged() to the StateFlow being collected
 to ensure we only handle the state change once.

donovanfm commented:
All looks good now. Thanks, Trevor!
